File: dsa-practice-ram/dsa-practice-ram/Arrays/DNF.cs
namespace dsa_practice_ram.Arrays
{
    internal class DNF
    {
        /*
         * 
         * [0...    ][1.....  ][unknown.....][2.......]
                      ^         ^          ^
	                  low      mid        high
        
            0 -> 0 - low - 1
            1 -> low - mid - 1
            2 -> high + 1 - n - 1
        */

        public void DutchNationalFlag()
        {
            int[] array = { 2, 1, 0, 1, 1, 2, 0, 0, 2, 1, 1, 1 };
            int low = 0;
            int mid = 0;
            int high = array.Length - 1;

            while (mid <= high)
            {
                int midValue = array[mid];

                if (midValue == 0)
                {
                    int temp = array[low];
                    array[low] = array[mid];
                    array[mid] = temp;

                    low++;
                    mid++;
                }
                else if (midValue == 1)
                {
                    mid++;
                }
                else
                {
                    int temp = array[high];
                    array[high] = array[mid];
                    array[mid] = temp;
                    high--;
                }
            }

            Console.WriteLine(string.Join(", ", array));
        }
    }
}

File: MyLearning/PythonPractice/dsa_practice/arrays/dnf.py
"""Dutch National Flag sorting."""

# C# DNF.DutchNationalFlag

def dutch_national_flag(array=None):
    """Sort an array of 0s, 1s, and 2s in-place.

    Problem: Reorder the array so all 0s come first, then 1s, then 2s.
    Approach: Three pointers (low, mid, high) partition the array in one pass.
    Time complexity: O(n)
    Space complexity: O(1) extra
    """
    if array is None:
        array = [2, 1, 0, 1, 1, 2, 0, 0, 2, 1, 1, 1]

    nums = list(array)
    low = 0
    mid = 0
    high = len(nums) - 1

    while mid <= high:
        mid_value = nums[mid]

        if mid_value == 0:
            nums[low], nums[mid] = nums[mid], nums[low]
            low += 1
            mid += 1
        elif mid_value == 1:
            mid += 1
        else:
            nums[high], nums[mid] = nums[mid], nums[high]
            high -= 1

    return nums
